apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -0,0 +1,53 @@
+package v1alpha1
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var log = logf.Log.WithName("baremetalhost-validation")
+
+func (host *BareMetalHost) Validate(old *BareMetalHost) error {
```

Comment:
AIUI putting implementing code in the API package is frowned-upon, I think mostly because it generally grows to start pulling in dependencies that other packages don't want when they just need to vendor the API. (I'm aware that we already have code in the API package, but people frowned upon that which is how we found out.)

I'd suggest we put this code in a new pkg/validation module.

Reply:
Agree. This is a good idea.

Reply:
Unfortunately, apis/ package has no access to pkg/ package. After checking other repos, it seems that validation should stay under apis/.

Reply:
I totally understand your concerns @zaneb and I spent some time about thinking and trying the problem. We have baremetalhost and baremetalhost_webhook under the apis directory and let's say we have a validation module under pkg.

apis/
* baremetalhost
* baremetalhost_webhook

pkg/
* validation

baremetalhost_webhook depends on validation
validation depends on baremetalhost

and this creates import cycle and not being compiled.

One option comes to my mind is move everything related to webhook under the pkg(I have not tried yet and not sure it is standard way or workaround).

I asked this problem on kubebuilder slack channel maybe we can get an idea about it.

Reply:
I've also have some concerns in putting more implementation code under the `/apis` - and in particular the webhooks.
I'm wondering if it's the case to review the current package organization to reduce the api surface, but it sounds something that could go out from the scope of the current PR.
Another approach could be for now to extract the validation code in a separate type, ie `BaremetalHostValidator`, so that it could be defined in a different package without the need of touching `apis/`

Reply:
I had a look through the controller-runtime code, and it seems that there are 3 requirements for the thing we pass to `NewWebhookManagedBy(mgr).For()`:

1. It must implement the [Validator](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/admission/validator.go#L30-L35) interface (i.e. have all of the `ValidateCreate()` &c. methods)
2. You must be able to unpack the BareMetalHost data into it
3. It must return the right group/version/kind when passed to [`GVKForObject()`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/apiutil/apimachinery.go#L75)

So I _think_ it would work to define a new type in another package:

```
type BareMetalHostValidator struct {
    v1alpha1.BareMetalHost
}

func (b *BareMetalHostValidator) ValidateCreate() error {
...
```

This is the way to go if it works.

Reply:
Thanks for your valuable comments. I think we all agree that webhook specific code will not be under `apis/`. 
My proposal is;
* First let me move current code under `pkg/` without creating a new `BaremetalHostValidator`. If it works, we don't have to create a struct for that. Everything under `pkg/` depends on `api/`
* If it does not work, it seems that `BaremetalHostValidator` might solve our problem.

Reply:
You can't add a new method to a type defined in another package in golang, so you will definitely have to create a new struct in order to move the code.

Reply:
Sorry, I forgot that webhook.Validator is BareMetalHost not new struct;
`var _ webhook.Validator = &BareMetalHost{}`

ok. I'll implement BaremetaHostValidator, thanks.

Reply:
This method did not work. In the setup register stage, it returns `kind BareMetalHostValidator not found` error(Because there is none). If `BareMetalHostValidator` is casted to `BareMetalHost`, then it returns `BareMetalHost` is not implemented as admission webhook interface. It is correct because `BareMetalHostValidator` has been implemented with that interface not `BareMetalHost`.

Since there isn't really any example about this usage for kubebuilder(and kubebuilder slack channel there was a comment about because time constraints it is designed to do simple validations, hence there is no such use case example). I started thinking that, maybe we can put these validations under the `apis/` folder. And we can have a chance to move it under `pkg/` when changing the folder structure as @andfasano stated.

Reply:
@zaneb I'd be concerned if you have 2 types returning the same gvk in the same process.
you are going to have really make sure the scheme is separate so the controller is getting the correct events. I'd suggest that controller-runtime *really* expects you to have the validation code in the api - and don't try to fight it...

Reply:
Btw, I also checked the controller-runtime, and IINW it appears that the dependency it's even [more strict](https://github.com/kubernetes-sigs/controller-runtime/blob/b704f447ea7c8f7059c6665143a4aa1f6da28328/pkg/webhook/admission/validator.go#L66), ie the decoded object from the admission request must implement the `Validator` interface.

Reply:
Yes, my point is if the underlying serialized object it's a BMH then there wouldn't be any way to workaround it (if not, there's still the registration point).
So to summarize it looks like there could be just two approaches:
* Put everything under `/apis`
* Refactor the packages structures